pallets/pallet-bonded-coins/src/benchmarking.rs

```diff
@@ -1 +1,855 @@
+	CollateralCurrenciesBalanceOf<T>: Into<U256> + TryFrom<U256>,
+	<T as Config>::CollateralCurrencies: Create<T::AccountId> ,
+	<T as Config>::Fungibles: InspectRoles<T::AccountId> + AccountTouch<FungiblesAssetIdOf<T>, AccountIdOf<T>>,
+	<T as Config>::DepositCurrency: Mutate<T::AccountId>,
+	<T as Config>::CollateralCurrencies: MutateFungibles<T::AccountId>,
+	AccountIdLookupOf<T>: From<T::AccountId>,
+)]
+mod benchmarks {
+	use frame_support::traits::OriginTrait;
+	use frame_support::traits::{
+		fungible::{Inspect, Mutate, MutateHold},
+		fungibles::{Destroy, Inspect as InspectFungibles, Mutate as MutateFungibles},
+	};
+	use frame_support::traits::{fungibles::Create, AccountTouch, EnsureOrigin, Get};
+	use sp_runtime::{traits::Zero, BoundedVec, SaturatedConversion};
+	use sp_std::ops::Mul;
+
+	use crate::{
+		curves::Curve,
+		mock::*,
+		types::{Locks, PoolManagingTeam, PoolStatus},
+		AccountIdLookupOf, AccountIdOf, CollateralAssetIdOf, CurveParameterInputOf, HoldReason, PoolDetailsOf, Pools,
+		TokenMetaOf,
+	};
+
+	use super::*;
+
+	// helper functions
+	// collateral currencies
+	fn create_collateral_asset<T: Config>(asset_id: CollateralAssetIdOf<T>)
+	where
+		<T as Config>::CollateralCurrencies: Create<T::AccountId>,
+	{
+		let pool_account = account("collateral_owner", 0, 0);
+		<T as Config>::CollateralCurrencies::create(asset_id, pool_account, false, 1u128.saturated_into())
+			.expect("Creating collateral asset should work");
+	}
+
+	fn calculate_default_collateral_asset_id<T: Config>() -> CollateralAssetIdOf<T> {
+		T::BenchmarkHelper::calculate_collateral_asset_id(u32::MAX)
+	}
+
+	fn create_default_collateral_asset<T: Config>() -> CollateralAssetIdOf<T>
+	where
+		<T as Config>::CollateralCurrencies: Create<T::AccountId>,
+	{
+		let collateral_id = calculate_default_collateral_asset_id::<T>();
+		create_collateral_asset::<T>(collateral_id.clone());
+		collateral_id
+	}
+
+	fn set_collateral_balance<T: Config>(asset_id: CollateralAssetIdOf<T>, who: &AccountIdOf<T>, amount: u128)
+	where
+		<T as Config>::CollateralCurrencies: MutateFungibles<T::AccountId>,
+	{
+		T::CollateralCurrencies::set_balance(asset_id, who, amount.saturated_into());
+	}
+
+	// bonded currencies
+
+	fn create_bonded_asset<T: Config>(asset_id: T::AssetId) {
+		let pool_account = account("bonded_owner", 0, 0);
+		<T as Config>::Fungibles::create(asset_id, pool_account, false, 1u128.saturated_into())
+			.expect("Creating bonded asset should work");
+	}
+
+	fn create_bonded_currencies_in_range<T: Config>(c: u32, is_destroying: bool) -> Vec<FungiblesAssetIdOf<T>> {
+		let mut asset_ids = Vec::new();
+		for i in 0..c {
+			let asset_id = T::BenchmarkHelper::calculate_bonded_asset_id(i);
+			asset_ids.push(asset_id.clone());
+			create_bonded_asset::<T>(asset_id.clone());
+			if is_destroying {
+				<T as Config>::Fungibles::start_destroy(asset_id, None).expect("Destroying should work");
+			}
+		}
```

Comment:
Asserts?

Reply:
This is quite challenging to verify. There is no trait available to access the state of an asset directly. However, I can introduce a storage alias to access the asset's storage and decode the result. The last byte of the encoded struct would reveal the state of the asset.

Direct access to the state is not possible because the properties are private

Reply:
But this would couple the benchmarking logic with the assets pallet, which is not necessarily true. If there's no way to verify this, just resolve this comment yourself.

Reply:
Yeah, true… I can introduce another function in the benchmarkhelper, to check if a currency is on a destroying state. That would be generic enough against all possible pallets, handling fungibles

Reply:
Like so: [03c19f0](https://github.com/KILTprotocol/kilt-node/pull/789/commits/03c19f008a7ecbe64c96f3108307225651a86034)

Reply:
No I personally think that is way overkill and risks to create a precedent where you require the trait to implement any test that you need in the benchmark. I would simply forget we had this conversation 😃

Reply:
Was feeling the same, but wanted to provide a possible solution. I reverted the changes in: [3617907](https://github.com/KILTprotocol/kilt-node/pull/789/commits/361790783098b2c8697cfc669c3588c2cc6e31cd)